File: 0872-Leaf-Similar-Trees/0872-Leaf-Similar-Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, vector<int> &v){
        if(root == NULL) return;
        //condition to check if it is a lead and adding the value into v
        if(root->left == NULL && root->right == NULL){
            v.push_back(root->val);
            return;
        }
        // for left traversal
        dfs(root->left,v);
        //for right traversal
        dfs(root->right,v);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int>tree1;
        vector<int>tree2;

        dfs(root1,tree1);
        dfs(root2,tree2);
        return tree1 == tree2;
    }
};

File: 0872-Leaf-Similar-Trees/0872-Leaf-Similar-Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root, v):
        if not root:
            return
        if not root.left and not root.right:
            v.append(root.val)
        self.dfs(root.left,v)
        self.dfs(root.right,v)
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        tree1 = []
        tree2 = []
        self.dfs(root1, tree1)
        self.dfs(root2, tree2)
        return tree1 == tree2
